qutebrowser/completion/models/miscmodels.py

```diff
@@ -138,3 +141,80 @@ def __init__(self, parent=None):
                     self.new_item(cat, name)
         except OSError:
             log.completion.exception("Failed to list sessions!")
+
+
+class TabCompletionModel(base.BaseCompletionModel):
+
+    """A model to complete on open tabs across all windows.
+
+    Used for switching the buffer command."""
+
+    # https://github.com/The-Compiler/qutebrowser/issues/545
+    # pylint: disable=abstract-method
+
+    #IDX_COLUMN = 0
+    URL_COLUMN = 1
+    TEXT_COLUMN = 2
+
+    COLUMN_WIDTHS = (6, 40, 54)
+    DUMB_SORT = Qt.DescendingOrder
+
+    def __init__(self, parent=None):
+        super().__init__(parent)
+
+        self.columns_to_filter = [self.URL_COLUMN, self.TEXT_COLUMN]
+
+        self._cats = []
+
+        for win_id in objreg.window_registry:
+            tabbed_browser = objreg.get('tabbed-browser', scope='window',
+                                    window=win_id)
+            for i in range(tabbed_browser.count()):
+                tab = tabbed_browser.widget(i)
+                tab.url_text_changed.connect(self.rebuild_cat)
+                tab.shutting_down.connect(self.on_tab_close)
+            tabbed_browser.new_tab.connect(self.rebuild_cat)
+        objreg.get("app").new_window.connect(self.on_new_window)
+        self.rebuild_cat()
+
+    # should be mainwindow.MainWindow but can't import that at module
+    # level because of import loops.
+    @pyqtSlot(object)
+    def on_new_window(self, window):
+        window.tabbed_browser.new_tab.connect(self.rebuild_cat)
+
+    @pyqtSlot()
+    def on_tab_close(self):
+        QTimer.singleShot(0, self.rebuild_cat)
+
+    @pyqtSlot(object)
+    def rebuild_cat(self, arg=None):
+        """Rebuild completion model from current tabs.
+
+        Very lazy method of keeping the model up to date. We could connect to
+        signals for new tab, tab url/title changed, tab close, tab moved and
+        make sure we handled background loads too ... but iterating over a
+        few/few dozen/few hundred tabs doesn't take very long at all.
+        """
+        # working with a tabbed_browser whilst shutting down throws some errors
```

Comment:
Okay I just removed that check and it still gets called when closing down
(from `on_tab_close()`) but no exceptions are thrown anymore. From memory it was
complaining about `tabbed_browser.widget(i)` or something because some of the Qt
stuff had been freed already. 

As to reproducing it, it wasn't a fatal error it was just stack traces being
printed but everything else worked fine. But I can't reproduce it now anyway.
And it is tested by exiting after initialising tab completion.